storage-node/src/services/webApi/controllers/filesApi.ts

```diff
@@ -34,35 +34,45 @@ export async function getFile(
   res: express.Response<unknown, AppConfig>,
   next: express.NextFunction
 ): Promise<void> {
+  const { id: dataObjectId } = req.params
   try {
-    await pinDataObjectIdToCache(req.params.id)
+    pinDataObjectIdToCache(dataObjectId)
+  } catch (err) {
+    res.status(404).send()
+    return
+  }
 
-    const dataObjectId = new BN(req.params.id)
+  try {
     const uploadsDir = res.locals.uploadsDir
-
-    const fullPath = path.resolve(uploadsDir, dataObjectId.toString())
-
+    const fullPath = path.resolve(uploadsDir, dataObjectId)
     const fileInfo = await getFileInfo(fullPath)
-    const fileStats = await fsPromises.stat(fullPath)
+
+    // verify ipfs has of file we are about to serve, just in-case it is corrupt
+    // ....
 
     const stream = send(req, fullPath)
 
     stream.on('headers', (res) => {
       // serve all files for download
       res.setHeader('Content-Disposition', 'inline')
       res.setHeader('Content-Type', fileInfo.mimeType)
-      res.setHeader('Content-Length', fileStats.size)
+      res.setHeader('Content-Length', fileInfo.size)
     })
 
     stream.on('error', (err) => {
       sendResponseWithError(res, next, err, 'files')
     })
 
+    stream.on('end', () => {
+      // we assume that if stream.pipe throws, then stream.on('end') will never fire
+      unpinDataObjectIdFromCache(dataObjectId)
+    })
+
     stream.pipe(res)
   } catch (err) {
     sendResponseWithError(res, next, err, 'files')
-  } finally {
-    await unpinDataObjectIdFromCache(req.params.id)
+    // we assume that if stream.pipe throws, then stream.on('end') will never fire
```

Comment:
The reason for the comment at both lines where we are invoking the unpin call is to explain that it is not a mistake (based on the assumption), and we should never actually get the result where unpin is invoked twice.
I suppose I could have a locally scoped function that will only even call `unpinDataObjectIdFromCache()` once and log a warning if it is every invoked multiple times as a signal that this assumption is false without causing a possible corrupt state or unpin throwing the assertion error.

Reply:
About this whole approach of pinning to prevent the cleanup service deleting a file being served, I may have overlooked that on linux file systems at least, it may not be necessary at all. I should have considered that when reviewing first PR that introduced the clean up service.

I do think we can safely get rid of this.

On top of that, I can't find the comment in the PR but I do remember at one point we were serving content from pending folder if objects was not yet moved to uploads, and I suggested that it was a bad idea as we didn't know what would happen if files was being being while being served.

I think we should re-introduce that. It will become more important when we move to object storage, if there is a delay in transferring/uploading the pending file from local file system (compared with what we do which is just renaming on the save disk volume).

What do you think?

Reply:
> On top of that, I can't find the comment in the PR but I do remember at one point we were serving content from pending folder if objects was not yet moved to uploads, and I suggested that it was a bad idea as we didn't know what would happen if files was being being while being served.

This is the PR where you suggested and mage the change https://github.com/Joystream/joystream/pull/4989

Reply:
Made unpinning less error prone and guaranteed to only happen once after pinning in getFile handler.
I read that end event on stream may/or may not fire if error event is fired, so there are potentially 3 places where we would need to unpin.. 

Done in [d5ec0ec](https://github.com/Joystream/joystream/pull/5016/commits/d5ec0ec0d15dc3271972c4c3b0885e124480b30b)

Reply:
> Made unpinning less error prone and guaranteed to only happen once after pinning in getFile handler.

Yeah, this is a good change actually.